src/mlio/image_reader.cxx

```diff
@@ -0,0 +1,413 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You
+ * may not use this file except in compliance with the License. A copy of
+ * the License is located at
+ *
+ *      http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is
+ * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+ * ANY KIND, either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+#include "mlio/image_reader.h"
+
+#ifdef MLIO_BUILD_IMAGE_READER
+
+#include <fmt/format.h>
+#include <opencv2/imgcodecs.hpp>
+#include <opencv2/imgproc/imgproc.hpp>
+
+#include "mlio/cpu_array.h"
+#include "mlio/data_type.h"
+#include "mlio/instance.h"
+#include "mlio/instance_batch.h"
+#include "mlio/intrusive_ptr.h"
+#include "mlio/logger.h"
+#include "mlio/parallel_data_reader.h"
+#include "mlio/record_readers/blob_record_reader.h"
+#include "mlio/record_readers/recordio_record_reader.h"
+#include "mlio/schema.h"
+
+namespace mlio {
+inline namespace v1 {
+
+image_reader::image_reader(data_reader_params rdr_prm,
+                           image_reader_params image_prm)
+    : parallel_data_reader{std::move(rdr_prm)}, params_{std::move(image_prm)}
+{
+    if (params_.image_dimensions.size() != image_dimensions_size_) {
+        throw std::invalid_argument{
+            "image_dimensions is a required parameter. "
+            "Dimensions of the output image must be entered in "
+            "(channels, height, width) format"};
+    }
+    bbh_ = effective_bad_batch_handling();
+    std::copy(params_.image_dimensions.begin(),
+              params_.image_dimensions.end(),
+              img_dims_.begin());
+}
+
+image_reader::~image_reader()
+{
+    stop();
+}
+
+intrusive_ptr<record_reader>
+image_reader::make_record_reader(data_store const &ds)
+{
+    switch (params_.img_frame) {
+    case image_frame::none:
+        return make_intrusive<blob_record_reader>(ds.open_read());
+    case image_frame::recordio:
+        return make_intrusive<detail::recordio_record_reader>(ds.open_read());
+    }
+    throw std::invalid_argument{
+        "The image_frame argument does not represent a valid value."};
+}
+
+intrusive_ptr<schema const>
+image_reader::infer_schema(std::optional<instance> const &)
+{
+    std::vector<attribute> attrs{};
+    // schema follows the NHWC convention
+    attrs.emplace_back(attribute_builder{"value",
```

Comment:
Minor. This can be reduced to:
```
    attrs.emplace_back("value",
                                       data_type::uint8,
                                       size_vector{params().batch_size,
                                       params_.image_dimensions[1],
                                       params_.image_dimensions[2],
                                       params_.image_dimensions[0]});
```